README.md
GLFW-CFFI
---------
[![Build Status](https://travis-ci.org/brianbruggeman/glfw-cffi.svg)](https://travis-ci.org/brianbruggeman/glfw-cffi)
[![PyPI version](https://img.shields.io/pypi/v/glfw-cffi.svg)](https://pypi.python.org/pypi/glfw-cffi)
[![Status](https://img.shields.io/pypi/status/glfw-cffi.svg)](https://pypi.python.org/pypi/glfw-cffi)
[![Python Compatibility](https://img.shields.io/pypi/pyversions/glfw-cffi.svg)](https://pypi.python.org/pypi/glfw-cffi)
[![Downloads](https://img.shields.io/pypi/dm/glfw-cffi.svg?period=week)](https://pypi.python.org/pypi/glfw-cffi)
[![Coverage Status](https://coveralls.io/repos/brianbruggeman/glfw-cffi/badge.svg?branch=develop&service=github)](https://coveralls.io/github/brianbruggeman/glfw-cffi?branch=develop)

A wrapper for GLFW3 using Python's CFFI.

## Motivation:

After being frustrated with the options available, I decided to roll my
own version of glfw3 using cffi for python.  This package inspects the
header file for glfw's api, and then creates python-wrapped functions.

For the most part, this should be transparent.  However, I have not
tested the package well enough to know where the bugs are.  So you may
use at your own risk.

Additionally, I have provided a straight copy of the GLFW3 api directly
available from the module.  So the module has python friendly functions
as well as the direct C-library calls.  Note that the latter API requires
more setup to use because you must manage the object conversion from
Python into C and back.

The goal for this package is that the user won't know they're using
a C library underneath and the package interface will feel completely
natural within Python.

## License:

This package is released as Apache 2.0 license.

However, at your option, you may apply any OSI approved free software
license you choose provided that you adhere to the free software license
chosen and additionally follow these criteria:

 a. list the author's name of this software as a contributor to your
    final product

 b. provide credit to your end user of your product or software without
    your end user asking for where you obtained your software

 c. notify the author of this software that you are using this software

 d. If you believe there can be some benefit in providing your changes
    upstream, you'll submit a change request.  While this criteria is
    completely optional, please consider not being a dick.

## Installation:

GLFW-CFFI uses GLFW3 and attempts to find the header file associated with your
specific library version to autogenerate the FFI interface.  So a version of
GLFW3 must be available during installation.  If a development version is
unavailable, then a version of the `glfw3.h` is included within the `glfw-cffi`
package itself.

### Installing via pip

Install via `pip install glfw-cffi`.

### Installing GLFW3

GLFW3 is available for several different platforms:

- Ubuntu/Debian: `sudo apt-get install -y libglfw3-dev`
- Fedora/Red Hat: `sudo yum install -y libglfw3-dev`
- Mac OS X with Homebrew: `brew install glfw3`
- Windows: There is are pre-compiled binaries available
  [64-bit Windows](https://github.com/glfw/glfw/releases/download/3.1.2/glfw-3.1.2.bin.WIN64.zip) or
  [32-bit Windows](https://github.com/glfw/glfw/releases/download/3.1.2/glfw-3.1.2.bin.WIN32.zip)

GLFW3 is relatively new, so some older installations of Linux may not have
`libglfw` directly available.  You may check out the [travis.yml](https://github.com/brianbruggeman/glfw-cffi/blob/master/.travis.yml#L34-L52)
file within our github repo for more information on setup on older systems.

#### A special note for installing GLFW3 on Windows

The current state requires that an environment variable, 'GLFW_LIBRARY', be set
and pointing to a compiled .dll found within a known path.  In addition,
glfw-cffi expects that a header file be present within an 'include' folder
within the same folder structure as the .dll.  So for example, if the
library binary were added to:

    C:\GLFW\lib\glfw3.dll

The python library, glfw-cffi, would search for a glfw3.h file within any of
these folders:

    C:\GLFW\lib\include
    C:\GLFW\include
    C:\include

When testing, we used the 32-bit binary and lib-mingw on a 64-bit Windows 10
system.

## Usage:

### Sample Usage:

This is the required code to produce a window on the screen:

    import glfw

    # Initialize glfw
    if not glfw.init():
        glfw.terminate()  # Cleans up if necessary
        raise RuntimeError('Could not initialize GLFW3')

    # Create window and set OpenGL Context
    win = glfw.create_window(title='Simple Window', width=640, height=480)
    glfw.make_context_current(win)

    # Main Loop
    while not glfw.window_should_close(win):
        glfw.swap_buffers(win)
        # To handle/process events use:
        glfw.poll_events()  # for continuous rendering (like in games)
        # or use:
        # glfw.wait_events()  # for on-event UIs (like an editing tool)

    # Proper shutdown
    glfw.terminate()

A more complex window example can be found within the examples folder on the github repo.

### Decorators

Extra decorators have been added to aid with developing a full user interface, including:

- keyboard handling
- mouse handling
- joystick handling
- window events
- text events
- path drop callbacks  (for drag and drop)
- error callbacks

Each decorator may be used with a standalone function or decorating a class method.
Examples of each type are found in the subsections below.  When decorating a class
method, use: @[staticmethod](https://docs.python.org/2/library/functions.html#staticmethod).

What follows is only the more commonly used.  Better documentation on callbacks can
be found on the [glfw website](http://www.glfw.org/docs/latest/).

#### Handling keyboard events

Keyboard events have a single decorator:

- keyboard event:  glfw.decorators.key_callback

Example:

    import glfw

    @glfw.decorators.key_callback
    def on_key(win, key, code, action, mods):
        '''Converts key into an event'''
        if key in [glfw.KEY_ESCAPE] and action in [glfw.PRESS]:
            glfw.set_window_should_close(win, gl.GL_TRUE)


In addition, helper functions have been added to convert data into strings:


    def display_data(key, action, mods):
        '''Converts keystroke into string data'''
        # Convert data
        key_action = glfw.get_key_string(key)
        action_string = glfw.get_action_string(action)
        mods_string = glfw.get_mod_string(mods)
        # Display data
        print('key: {key} -> "{string}"'.format(key=key, string=key_string))
        print('action: {action} -> "{string}"'.format(action=action, string=action_string))
        print('mods: {mods} -> "{string}"'.format(mods=mods, string=mods_string))


Finally, sometimes keystroke handling may make sense to be included within a class.


    import glfw
    from OpenGL import GL as gl

    class Foo(object):

        @staticmethod
        @glfw.decorators.key_callback
        def on_key(win, key, code, action, mods):
            '''Handles a key event'''
            if key in [glfw.KEY_ESCAPE] and action in [glfw.PRESS]:
                glfw.set_window_should_close(win, gl.GL_TRUE)
            # Display what just happened
            key = glfw.get_key_string(key)
            amapping = {'press': '+', 'release': '-', 'repeat': '*'}
            action = amapping.get(glfw.get_action_string(action))
            mods = glfw.get_mod_string(mods)
            string = '{}|{}'.format(action[0], '+'.join(str(_) for _ in (mods, key) if _))
            print(string)


#### Handling mouse events

Mouse events have three decorators:

- mouse button click:  glfw.decorators.mouse_button_callback
- mouse wheel/scroll: glfw.decorators.scroll_callback
- mouse movement: glfw.decorators.cursor_pos_callback

Example:

    import glfw
    from OpenGL import GL as gl


    class Foo(object):

        @staticmethod
        @glfw.decorators.mouse_button_callback
        def on_mouse_button(win, button, action, mods):
            '''Handles a mouse button event'''
            # Not used here, but having the position where the mouse was at the
            #  time of the click can be useful.
            position = glfw.get_cursor_pos(win)
            # Handle button
            if button in [glfw.MOUSE_BUTTON_1] and action in [glfw.PRESS]:
                glfw.set_window_should_close(win, gl.GL_TRUE)
            # Display what just happened
            button = glfw.get_mouse_button_string(button)
            amapping = {'press': '+', 'release': '-', 'repeat': '*'}
            action = amapping.get(glfw.get_action_string(action))
            mods = glfw.get_mod_string(mods)
            position = '({:>.0f}, {:>.0f})'.format(*position)
            string = '{} {}|{}'.format(position, action[0], '+'.join(str(_) for _ in (mods, button) if _))
            print(string)

        @staticmethod
        @glfw.decorators.scroll_callback
        def on_mouse_scroll(win, x_offset, y_offset):
            '''Handles a mouse scroll/wheel event'''

        @staticmethod
        @glfw.decorators.cursor_pos_callback
        def on_mouse_move(win, x_offset, y_offset):
            '''Handles a mouse move event'''



#### Handling window events

There are other available decorators that handle window events.

##### Gaining and Losing Focus

Windows client areas may gain or lose focus and an event is
triggered each time.

- client focus: glfw.decorators.cursor_enter_callback
- window focus: glfw.decorators.window_focus_callback

Example:

    import glfw


    class Foo(object):

        @staticmethod
        @glfw.decorators.cursor_enter_callback
        def on_enter(win, status):
          '''Handles focus event for a window client area

          status is a boolean:  True for focused and False for unfocused
          '''

        @staticmethod
        @glfw.decorators.window_focus_callback
        def on_enter(win, status):
          '''Handles focus event for a window

          status is a boolean:  True for focused and False for unfocused
          '''

##### Resizing

Windows may be resized.

- resize: glfw.decorators.window_size_callback

Example:

    import glfw


    class Foo(object):

        @staticmethod
        @glfw.decorators.window_size_callback
        def on_enter(win, width, height):
          '''Handles resize event'''


## Examples:

More examples can be within the github repo under the [examples/](https://github.com/brianbruggeman/glfw-cffi/tree/develop/examples) folder.

Some of the examples require more packages to be installed:

- [docopt](https://pypi.python.org/pypi/docopt):  Creates beautiful command-line interfaces
- [numpy](https://pypi.python.org/pypi/numpy):  is a general-purpose array-processing package designed to efficiently manipulate large multi-dimensional arrays of arbitrary records without sacrificing too much speed for small multi-dimensional arrays
- [freetype-py](https://pypi.python.org/pypi/freetype-py/): Freetype python provides bindings for the FreeType library. Only the high-level API is bound.


## Contributions:

Contributions are welcome. When opening a PR, please keep the following guidelines in mind:

- Before implementing, please open an issue for discussion.
- Make sure you have tests for the new logic.
- Make sure your code passes `flake8`
- Add yourself to contributors at `README.md` and/or  your contributions.

## Contributors

* [Brian Bruggeman](https://github.com/brianbruggeman) - Originator


CHANGES.rst
0.3.0
-----
* glfw.init and glfw.terminate now setup during `import glfw`
* Updated package structure
* Added tests, test fixtures and increased coverage
* Updated travis build to use GLFW 3.2.1
* Updated tox work flows
* Updated examples with new package structure
* Adds docker


0.2.0
-----

* Fixes issues with GLFW Vulkan support (Vulkan is not supported)
* Adds docs


examples/Window.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
Showcases how GLFW-CFFI might be wrapped in a higher level Window class.

Usage:
    Window.py [options]

Options:
    -h --help           This message
    -q --quiet          Less messages
    -v --verbose        More messages
    -w --width WIDTH    Width of window [default: 640]
    -H --height HEIGHT  Height of window [default: 480]
    -t --title TITLE    Title of window [default: GLFW Example]
'''
from __future__ import print_function

import logging
import threading

import OpenGL
# Turn off ERROR_CHECKING to improve OpenGL performance
#  This must be run before glfw is imported
OpenGL.ERROR_CHECKING = True

import glfw
# GLFW comes with a replacement for OpenGL which uses OpenGL but renames
#  the functions so they are snake_case and drops the GL_ prefix
#  from enumerations
from glfw import gl


log = logging.getLogger('Window')


class Window(object):
    '''Wraps GLFW functions into a convenient package

    Args:
        title(str): Name of the window
        height(int): height of the window
        width(int): width of the window

    >>> win = Window(title='Example', width=1080, height=720)
    >>> win.loop()
    '''
    registry = {}

    @property
    def width(self):
        '''Window width'''
        width, height = glfw.get_window_size(self.win)
        return width

    @property
    def height(self):
        '''Window height'''
        width, height = glfw.get_window_size(self.win)
        return height

    @property
    def fb_width(self):
        '''Framebuffer width'''
        fb_width, fb_height = glfw.get_framebuffer_size(self.win)
        return fb_width

    @property
    def fb_height(self):
        '''Framebuffer height'''
        fb_width, fb_height = glfw.get_framebuffer_size(self.win)
        return fb_height

    def __init__(self, title='GLFW Example', height=None, width=None, monitor=None):
        # Determine available major/minor compatibility
        #  This contains init and terminate logic for glfw, so it must be run first
        major, minor = self.get_opengl_version()
        self.title = title

        # Lock is for thread aware Windows and opening, closing and garbage
        #  collection
        self.lock = threading.Lock()

        # Hinting must be run before window creation
        glfw.core.window_hint(glfw.CONTEXT_VERSION_MAJOR, major)
        glfw.core.window_hint(glfw.CONTEXT_VERSION_MINOR, minor)
        glfw.core.window_hint(glfw.OPENGL_PROFILE, glfw.OPENGL_CORE_PROFILE)
        glfw.core.window_hint(glfw.OPENGL_FORWARD_COMPAT, gl.TRUE)

        # Generate window
        if monitor is None:
            count = glfw.ffi.new('int *')
            monitor_list = glfw.core.get_monitors(count)
            monitor_list = [monitor_list[index] for index in range(count[0])]

            # handle = glfw.get_primary_monitor()
            # monitor = handle
        if height is None and width is None:
            if not isinstance(monitor, glfw.Monitor):
                monitor = glfw.Monitor(monitor)
            width, height = monitor.resolution
        self.win = glfw.create_window(height=height, width=width, title=title)  # , monitor=monitor)
        Window.registry[self.win] = self

        # Setup window callbacks: Must be run after creating an OpenGL window
        self.setup_callbacks()

        # Set context
        glfw.core.make_context_current(self.win)
        self.init()

    def __del__(self):
        '''Removes the glfw window'''
        if hasattr(self, 'win'):
            glfw.core.set_window_should_close(self.win, gl.TRUE)
            # Wait for loop to end
            self.lock.acquire()
            glfw.destroy_window(self.win)
            self.lock.release()

    def get_opengl_version(self):
        '''Contains logic to determine opengl version.

        Only run this within initialization before running standard window code
        '''
        # Determine available major/minor compatibility
        #  This contains init and terminate logic for glfw, so it must be run first
        ffi = glfw.ffi
        opengl_version = None
        versions = [
            (4, 5), (4, 4), (4, 3), (4, 2), (4, 1), (4, 0),
            (3, 3), (3, 2), (3, 1), (3, 0),
            (2, 1), (2, 0),
            (1, 5), (1, 4), (1, 3), (1, 2), (1, 1), (1, 0),
        ]
        farg = ffi.new('char []', bytes(''.encode('utf-8')))
        title = farg

        if not glfw.core.init():
            glfw.terminate()
            raise RuntimeError('Could not initialize GLFW')

        for major, minor in versions:
            try:
                glfw.window_hint(glfw.CONTEXT_VERSION_MAJOR, major)
                glfw.window_hint(glfw.CONTEXT_VERSION_MINOR, minor)
                glfw.window_hint(glfw.OPENGL_PROFILE, glfw.OPENGL_CORE_PROFILE)
                glfw.window_hint(glfw.OPENGL_FORWARD_COMPAT, gl.GL_TRUE)
                glfw.window_hint(glfw.VISIBLE, gl.GL_FALSE)
                glfw.window_hint(glfw.FOCUSED, gl.GL_FALSE)
                window = glfw.core.create_window(1, 1, title, ffi.NULL, ffi.NULL)
                if window != ffi.NULL:
                    glfw.destroy_window(window)
                    if opengl_version is None:
                        opengl_version = (major, minor)
            except Exception as e:
                import traceback as tb
                for line in tb.format_exc(e).split('\n'):
                    log.error(line)

        glfw.terminate()
        return opengl_version

    def init(self):
        '''Scene initialization'''
        gl.clear(gl.COLOR_BUFFER_BIT)

    def render(self):
        '''Empty scene'''

    def loop(self):
        '''Simplified loop'''
        self.lock.acquire()
        while not glfw.window_should_close(self.win):
            self.render()
            self.handle_buffers_and_events()
        self.lock.release()

    def handle_buffers_and_events(self):
        glfw.core.swap_buffers(self.win)
        glfw.core.poll_events()

    def setup_callbacks(self):
        '''Creates glfw callbacks for this window'''
        glfw.core.set_char_callback(self.win, self.on_char)
        glfw.core.set_char_mods_callback(self.win, self.on_char_mods)
        glfw.core.set_cursor_enter_callback(self.win, self.on_cursor_focus)
        glfw.core.set_cursor_pos_callback(self.win, self.on_mouse_move)
        glfw.core.set_error_callback(self.on_error)
        glfw.core.set_key_callback(self.win, self.on_key)
        glfw.core.set_mouse_button_callback(self.win, self.on_mouse_button)
        glfw.core.set_scroll_callback(self.win, self.on_scroll)
        glfw.core.set_window_refresh_callback(self.win, self.on_window_refresh)
        glfw.core.set_window_size_callback(self.win, self.on_window_resize)
        glfw.core.set_window_pos_callback(self.win, self.on_window_move)
        glfw.core.set_window_close_callback(self.win, self.on_window_close)
        glfw.core.set_window_focus_callback(self.win, self.on_window_focus)
        glfw.core.set_window_iconify_callback(self.win, self.on_window_minimize)
        glfw.core.set_drop_callback(self.win, self.on_file_drag_and_drop)
        glfw.core.set_framebuffer_size_callback(self.win, self.on_framebuffer_resize)
        glfw.core.set_monitor_callback(self.on_monitor)

    @staticmethod
    def cdata_to_pystring(cdata):
        '''Converts char * cdata into a python string'''
        return glfw.ffi_string(cdata)

    @staticmethod
    @glfw.decorators.char_callback
    def on_char(win, codepoint):
        '''Handles unicode char callback

        This is useful for handling simple character strokes and text input
        '''
        log.debug('{}'.format(chr(codepoint)))

    @staticmethod
    @glfw.decorators.char_mods_callback
    def on_char_mods(win, codepoint, mods):
        '''Handles unicode char callback /w mods

        This is useful for handling simple character strokes and text input
        with modifiers
        '''
        msg = '+'.join('{}'.format(_) for _ in (chr(codepoint), glfw.get_mod_string(mods)))
        log.debug('{}'.format(msg))

    @staticmethod
    @glfw.decorators.cursor_enter_callback
    def on_cursor_focus(win, state):
        '''Cursor position upon entering client area'''
        state = 'Gained Focus' if state else 'Lost Focus'
        log.debug('Mouse Focus: {state}'.format(state=state))

    @staticmethod
    @glfw.decorators.drop_callback
    def on_file_drag_and_drop(win, count, paths):
        '''Handles drag and drop of file paths'''
        for some_path in range(count):
            path = Window.cdata_to_pystring(paths[some_path])
            log.debug('Drag and dropped file: {}'.format(path))

    @staticmethod
    @glfw.decorators.error_callback
    def on_error(code, message):
        '''Handles an error callback event'''
        error_message = Window.cdata_to_pystring(message)
        message = '{}: {}'.format(code, error_message)
        log.error(message)

    @staticmethod
    @glfw.decorators.framebuffer_size_callback
    def on_framebuffer_resize(win, width, height):
        '''Handles a framebuffer resize event'''
        log.debug('New Framebuffer Size: ({width}, {height})'.format(width=width, height=height))

    @staticmethod
    @glfw.decorators.key_callback
    def on_key(win, key, code, action, mods):
        '''Handles a key event'''
        if key in [glfw.KEY_ESCAPE] and action in [glfw.PRESS]:
            glfw.set_window_should_close(win, gl.TRUE)
        # Display what just happened
        key = glfw.get_key_string(key)
        amapping = {'press': '+', 'release': '-', 'repeat': '*'}
        action = amapping.get(glfw.get_action_string(action))
        mods = glfw.get_mod_string(mods)
        string = '{}|{}'.format(action[0], '+'.join(str(_) for _ in (mods, key) if _))
        log.debug(string)

    @staticmethod
    @glfw.decorators.monitor_callback
    def on_monitor(monitor, event=None):
        '''Handles monitor connect and disconnect'''
        state = 'Connected' if event else 'Disconnected'
        log.debug('Monitor {event}: {monitor}'.format(event=event, monitor=state))

    @staticmethod
    @glfw.decorators.mouse_button_callback
    def on_mouse_button(win, button, action, mods):
        '''Handles a mouse button event'''
        # Not used here, but having the position where the mouse was at the
        #  time of the click can be useful.
        position = glfw.get_cursor_pos(win)
        # Handle button
        if button in [glfw.MOUSE_BUTTON_1] and action in [glfw.PRESS]:
            glfw.set_window_should_close(win, gl.TRUE)
        # Display what just happened
        button = glfw.get_mouse_button_string(button)
        amapping = {'press': '+', 'release': '-', 'repeat': '*'}
        action = amapping.get(glfw.get_action_string(action))
        mods = glfw.get_mod_string(mods)
        position = '({:>.0f}, {:>.0f})'.format(*position)
        string = '{} {}|{}'.format(position, action[0], '+'.join(str(_) for _ in (mods, button) if _))
        log.debug(string)

    @staticmethod
    @glfw.decorators.cursor_pos_callback
    def on_mouse_move(win, x, y):
        '''Mouse movement handler'''
        log.debug('Mouse position: ({x}, {y})'.format(x=x, y=y))

    @staticmethod
    @glfw.decorators.scroll_callback
    def on_scroll(win, x, y):
        '''Scrollback handler'''
        log.debug('Mouse scroll: ({x}, {y})'.format(x=x, y=y))

    @staticmethod
    @glfw.decorators.window_close_callback
    def on_window_close(win):
        '''Handles a window close event'''
        win = Window.registry.get(win)
        if win:
            win = win.title
        log.debug('Window closed: ({win})'.format(win=win))

    @staticmethod
    @glfw.decorators.window_focus_callback
    def on_window_focus(win, state):
        ''''''
        state = 'Regained Focus' if state else 'Lost Focus'
        log.debug('Window: {state}'.format(state=state))

    @staticmethod
    @glfw.decorators.window_iconify_callback
    def on_window_minimize(win, state):
        '''Handles window minimization/restore events'''
        state = 'Minimized' if state else 'Restored'
        log.debug('Window {state}: {win}'.format(state=state, win=win))

    @staticmethod
    @glfw.decorators.window_pos_callback
    def on_window_move(win, x, y):
        '''Handles window move event'''
        log.debug('New window position: ({x}, {y})'.format(x=x, y=y))

    @staticmethod
    @glfw.decorators.window_refresh_callback
    def on_window_refresh(win):
        '''Window refresh handler

        Called periodically or during window resize events'''
        log.debug('Refreshed window: {win}'.format(win=win))

    @staticmethod
    @glfw.decorators.window_size_callback
    def on_window_resize(win, width, height):
        '''Window resize handler'''
        log.debug('Window size: {width}x{height}'.format(width=width, height=height))


def main(**options):
    width = int(options.get('width'))
    height = int(options.get('height'))
    title = options.get('title')
    window = Window(title=title, height=height, width=width)
    window.loop()


if __name__ == '__main__':
    from docopt import docopt

    def fix(option):
        option = option.lstrip('--')  # --optional-arg -> optional-arg
        option = option.lstrip('<').rstrip('>')  # <positional-arg> -> positional-arg
        option = option.replace('-', '_')  # hyphen-arg -> method_parameter
        return option

    options = {fix(k): v for k, v in docopt(__doc__).items()}
    if options.get('quiet'):
        logging.basicConfig(level=logging.WARNING)
    elif options.get('verbose'):
        logging.basicConfig(level=logging.DEBUG)
    else:
        logging.basicConfig(level=logging.INFO)

    main(**options)
